File: observer/ts/iot-light-ex/iot-light.ts
import { EventEmitter } from 'events';

// 抽象主題 (Subject) - 使用 EventEmitter 簡化實現
// 在 TypeScript 中，通常會使用介面或抽象類別來定義主題和觀察者的契約。
// 這裡為了簡化，我們讓 StreetLightController 繼承 EventEmitter，
// 這樣它就自動擁有了 `on`, `off`, `emit` 等方法來管理事件和通知觀察者。
enum LightStatus {
  ON, // 0
  OFF, // 1
  FAULT, // 2
}

interface StreetLightState {
    lightId: string;
    status: 'OFF' | 'ON' | 'FAULT';
    brightness: number; // 0-100
}

abstract class StreetLightController extends EventEmitter {
    protected _lightId: string;
    protected _status: 'OFF' | 'ON' | 'FAULT';
    protected _brightness: number; // 0-100

    constructor(lightId: string) {
        super();
        this._lightId = lightId;
        this._status = 'OFF';
        this._brightness = 0;
    }

    // 在 EventEmitter 的基礎上，我們不需要手動管理 observers 列表和 notify 方法。
    // 而是通過發射事件來通知觀察者。
    // attach 和 detach 方法可以被 `on` 和 `off` 方法替代。

    get lightId(): string {
        return this._lightId;
    }

    get status(): 'OFF' | 'ON' | 'FAULT' {
        return this._status;
    }

    get brightness(): number {
        return this._brightness;
    }

    protected notifyObservers(): void {
        console.log(`\n[路燈 ${this._lightId}] 狀態更新，通知所有監控服務...`);
        // 發射一個 'update' 事件，並傳遞當前路燈的狀態
        this.emit('update', { 
            lightId: this._lightId,
            status: this._status,
            brightness: this._brightness
        } as StreetLightState);
    }

    abstract turnOn(brightness?: number): void;
    abstract turnOff(): void;
    abstract setBrightness(brightness: number): void;
    abstract simulateFault(): void;
}

// 抽象觀察者 (Observer) - 使用介面定義契約
interface MonitoringSystem {
    name: string;
    update(state: StreetLightState): void;
}

// 具體主題 (Concrete Subject)
class ConcreteStreetLightController extends StreetLightController {
    constructor(lightId: string) {
        super(lightId);
    }

    turnOn(brightness: number = 100): void {
        const newBrightness = Math.max(0, Math.min(100, brightness));
        if (this._status !== 'ON' || this._brightness !== newBrightness) {
            this._status = 'ON';
            this._brightness = newBrightness;
            console.log(`[路燈 ${this._lightId}] 狀態變更: 開啟, 亮度 ${this._brightness}%`);
            this.notifyObservers();
        }
    }

    turnOff(): void {
        if (this._status !== 'OFF') {
            this._status = 'OFF';
            this._brightness = 0;
            console.log(`[路燈 ${this._lightId}] 狀態變更: 關閉`);
            this.notifyObservers();
        }
    }

    setBrightness(brightness: number): void {
        const newBrightness = Math.max(0, Math.min(100, brightness));
        if (this._status === 'ON' && this._brightness !== newBrightness) {
            this._brightness = newBrightness;
            console.log(`[路燈 ${this._lightId}] 亮度變更: ${this._brightness}%`);
            this.notifyObservers();
        } else if (this._status === 'OFF') {
            console.log(`[路燈 ${this._lightId}] 路燈已關閉，無法設定亮度。`);
        } else if (this._status === 'FAULT') {
            console.log(`[路燈 ${this._lightId}] 路燈故障中，無法設定亮度。`);
        }
    }

    simulateFault(): void {
        if (this._status !== 'FAULT') {
            this._status = 'FAULT';
            this._brightness = 0;
            console.log(`[路燈 ${this._lightId}] 狀態變更: 故障！`);
            this.notifyObservers();
        }
    }
}

// 具體觀察者 (Concrete Observer)
class DashboardDisplay implements MonitoringSystem {
    name: string;

    constructor(name: string = "儀表板顯示") {
        this.name = name;
    }

    update(state: StreetLightState): void {
        console.log(`  [${this.name}] 收到路燈 ${state.lightId} 更新: 狀態=${state.status}, 亮度=${state.brightness}%`);
        if (state.status === 'FAULT') {
            console.log(`  [${this.name}] 警告: 路燈 ${state.lightId} 發生故障！`);
        }
    }
}

class AlertSystem implements MonitoringSystem {
    name: string;

    constructor(name: string = "警報系統") {
        this.name = name;
    }

    update(state: StreetLightState): void {
        if (state.status === 'FAULT') {
            console.log(`  [${this.name}] !!! 緊急警報 !!! 路燈 ${state.lightId} 故障，請立即處理！`);
        } else if (state.status === 'OFF' && state.brightness === 0) {
            console.log(`  [${this.name}] 通知: 路燈 ${state.lightId} 已關閉。`);
        }
    }
}

class LoggingService implements MonitoringSystem {
    name: string;

    constructor(name: string = "日誌服務") {
        this.name = name;
    }

    update(state: StreetLightState): void {
        console.log(`  [${this.name}] 記錄: 路燈 ${state.lightId} 狀態從 ${state.status} 變更，亮度 ${state.brightness}%`);
    }
}

// --- 範例使用 --- //
async function main() {
    console.log("===== IoT 路燈控制器監控平台 Observer 模式範例 (TypeScript) =====\n");

    // 1. 創建多個路燈控制器 (主題)
    const lightControllerA = new ConcreteStreetLightController("SL-001");
    const lightControllerB = new ConcreteStreetLightController("SL-002");
    const lightControllerC = new ConcreteStreetLightController("SL-003");

    // 2. 創建多個監控服務 (觀察者)
    const dashboard = new DashboardDisplay();
    const alertSystem = new AlertSystem();
    const loggingService = new LoggingService();

    // 3. 監控服務訂閱感興趣的路燈
    // 在 TypeScript 中，我們使用 EventEmitter 的 `on` 方法來訂閱事件。
    // 這裡的 'update' 是我們自定義的事件名稱。

    // 儀表板監控所有路燈
    lightControllerA.on('update', (state: StreetLightState) => dashboard.update(state));
    console.log(`[路燈 ${lightControllerA.lightId}] 添加監控服務: ${dashboard.name}`);
    lightControllerB.on('update', (state: StreetLightState) => dashboard.update(state));
    console.log(`[路燈 ${lightControllerB.lightId}] 添加監控服務: ${dashboard.name}`);
    lightControllerC.on('update', (state: StreetLightState) => dashboard.update(state));
    console.log(`[路燈 ${lightControllerC.lightId}] 添加監控服務: ${dashboard.name}`);

    // 警報系統只監控 SL-001 和 SL-003 的故障
    const slc1AlertListener = (state: StreetLightState) => alertSystem.update(state);
    lightControllerA.on('update', slc1AlertListener);
    console.log(`[路燈 ${lightControllerA.lightId}] 添加監控服務: ${alertSystem.name}`);
    const slc3AlertListener = (state: StreetLightState) => alertSystem.update(state);
    lightControllerC.on('update', slc3AlertListener);
    console.log(`[路燈 ${lightControllerC.lightId}] 添加監控服務: ${alertSystem.name}`);

    // 日誌服務監控所有路燈
    lightControllerA.on('update', (state: StreetLightState) => loggingService.update(state));
    console.log(`[路燈 ${lightControllerA.lightId}] 添加監控服務: ${loggingService.name}`);
    lightControllerB.on('update', (state: StreetLightState) => loggingService.update(state));
    console.log(`[路燈 ${lightControllerB.lightId}] 添加監控服務: ${loggingService.name}`);
    lightControllerC.on('update', (state: StreetLightState) => loggingService.update(state));
    console.log(`[路燈 ${lightControllerC.lightId}] 添加監控服務: ${loggingService.name}`);

    console.log("\n--- 模擬路燈操作與狀態變化 ---");

    console.log("\n--- 路燈 SL-001 開啟 ---");
    lightControllerA.turnOn(80);
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n--- 路燈 SL-002 開啟並調整亮度 ---");
    lightControllerB.turnOn(50);
    await new Promise(resolve => setTimeout(resolve, 500));
    lightControllerB.setBrightness(75);
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n--- 路燈 SL-003 模擬故障 ---");
    lightControllerC.simulateFault();
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n--- 路燈 SL-001 調整亮度 ---");
    lightControllerA.setBrightness(100);
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n--- 路燈 SL-002 關閉 ---");
    lightControllerB.turnOff();
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n--- 警報系統不再監控 SL-003 ---");
    // 為了正確移除，需要使用保存的函數引用
    lightControllerC.off('update', slc3AlertListener);
    console.log(`[路燈 ${lightControllerC.lightId}] 移除監控服務: ${alertSystem.name}`);

    console.log("\n--- 路燈 SL-003 故障排除 (假設修復後重新開啟) ---");
    lightControllerC.turnOn(90);
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log("\n===== 範例結束 =====");
}

main();

File: observer/py/iot-light-ex/iot-light.py
from abc import ABC, abstractmethod
import time
import random

# 抽象主題 (Subject)
class StreetLightController(ABC):
    """路燈控制器抽象主題，定義了管理觀察者和通知觀察者的方法。"""
    def __init__(self, light_id):
        self._light_id = light_id
        self._status = "OFF"  # OFF, ON, FAULT
        self._brightness = 0  # 0-100
        self._observers = []

    def attach(self, observer):
        """添加一個觀察者。"""
        if observer not in self._observers:
            self._observers.append(observer)
            print(f"[路燈 {self._light_id}] 添加監控服務: {observer.name}")

    def detach(self, observer):
        """移除一個觀察者。"""
        try:
            self._observers.remove(observer)
            print(f"[路燈 {self._light_id}] 移除監控服務: {observer.name}")
        except ValueError:
            pass

    def notify(self):
        """通知所有註冊的觀察者。"""
        print(f"\n[路燈 {self._light_id}] 狀態更新，通知所有監控服務...")
        for observer in self._observers:
            observer.update(self)

    @property
    def light_id(self):
        return self._light_id

    @property
    def status(self):
        return self._status

    @property
    def brightness(self):
        return self._brightness

    @abstractmethod
    def turn_on(self, brightness=100):
        """開啟路燈。"""
        pass

    @abstractmethod
    def turn_off(self):
        """關閉路燈。"""
        pass

    @abstractmethod
    def set_brightness(self, brightness):
        """設定路燈亮度。"""
        pass

    @abstractmethod
    def simulate_fault(self):
        """模擬路燈故障。"""
        pass

# 抽象觀察者 (Observer)
class MonitoringSystem(ABC):
    """監控系統抽象觀察者，定義了接收更新的方法。"""
    def __init__(self, name):
        self._name = name

    @property
    def name(self):
        return self._name

    @abstractmethod
    def update(self, subject: StreetLightController):
        """當主題狀態改變時被呼叫。"""
        pass

# 具體主題 (Concrete Subject)
class ConcreteStreetLightController(StreetLightController):
    """具體路燈控制器，管理其狀態並在狀態變化時通知觀察者。"""
    def turn_on(self, brightness=100):
        if self._status != "ON" or self._brightness != brightness:
            self._status = "ON"
            self._brightness = max(0, min(100, brightness))
            print(f"[路燈 {self._light_id}] 狀態變更: 開啟, 亮度 {self._brightness}%")
            self.notify()

    def turn_off(self):
        if self._status != "OFF":
            self._status = "OFF"
            self._brightness = 0
            print(f"[路燈 {self._light_id}] 狀態變更: 關閉")
            self.notify()

    def set_brightness(self, brightness):
        new_brightness = max(0, min(100, brightness))
        if self._status == "ON" and self._brightness != new_brightness:
            self._brightness = new_brightness
            print(f"[路燈 {self._light_id}] 亮度變更: {self._brightness}%")
            self.notify()
        elif self._status == "OFF":
            print(f"[路燈 {self._light_id}] 路燈已關閉，無法設定亮度。")
        elif self._status == "FAULT":
            print(f"[路燈 {self._light_id}] 路燈故障中，無法設定亮度。")

    def simulate_fault(self):
        if self._status != "FAULT":
            self._status = "FAULT"
            self._brightness = 0
            print(f"[路燈 {self._light_id}] 狀態變更: 故障！")
            self.notify()

# 具體觀察者 (Concrete Observer)
class DashboardDisplay(MonitoringSystem):
    """儀表板顯示服務，顯示路燈的即時狀態。"""
    def __init__(self, name="儀表板顯示"): # Assign a default name
        super().__init__(name)

    def update(self, subject: StreetLightController):
        print(f"  [{self.name}] 收到路燈 {subject.light_id} 更新: 狀態={subject.status}, 亮度={subject.brightness}%")
        if subject.status == "FAULT":
            print(f"  [{self.name}] 警告: 路燈 {subject.light_id} 發生故障！")

class AlertSystem(MonitoringSystem):
    """警報系統，在路燈發生故障時發送警報。"""
    def __init__(self, name="警報系統"): # Assign a default name
        super().__init__(name)

    def update(self, subject: StreetLightController):
        if subject.status == "FAULT":
            print(f"  [{self.name}] !!! 緊急警報 !!! 路燈 {subject.light_id} 故障，請立即處理！")
        elif subject.status == "OFF" and subject.brightness == 0:
            print(f"  [{self.name}] 通知: 路燈 {subject.light_id} 已關閉。")

class LoggingService(MonitoringSystem):
    """日誌服務，記錄路燈的所有狀態變化。"""
    def __init__(self, name="日誌服務"): # Assign a default name
        super().__init__(name)

    def update(self, subject: StreetLightController):
        print(f"  [{self.name}] 記錄: 路燈 {subject.light_id} 狀態從 {subject.status} 變更，亮度 {subject.brightness}%")

# --- 範例使用 --- #
if __name__ == "__main__":
    print("===== IoT 路燈控制器監控平台 Observer 模式範例 =====\n")

    # 1. 創建多個路燈控制器 (主題)
    light_controller_A = ConcreteStreetLightController("SL-001")
    light_controller_B = ConcreteStreetLightController("SL-002")
    light_controller_C = ConcreteStreetLightController("SL-003")

    # 2. 創建多個監控服務 (觀察者)
    dashboard = DashboardDisplay()
    alert_system = AlertSystem()
    logging_service = LoggingService()

    # 3. 監控服務訂閱感興趣的路燈
    # 儀表板監控所有路燈
    light_controller_A.attach(dashboard)
    light_controller_B.attach(dashboard)
    light_controller_C.attach(dashboard)

    # 警報系統只監控 SL-001 和 SL-003 的故障
    light_controller_A.attach(alert_system)
    light_controller_C.attach(alert_system)

    # 日誌服務監控所有路燈
    light_controller_A.attach(logging_service)
    light_controller_B.attach(logging_service)
    light_controller_C.attach(logging_service)

    print("\n--- 模擬路燈操作與狀態變化 ---")

    print("\n--- 路燈 SL-001 開啟 ---")
    light_controller_A.turn_on(80)
    time.sleep(0.5)

    print("\n--- 路燈 SL-002 開啟並調整亮度 ---")
    light_controller_B.turn_on(50)
    time.sleep(0.5)
    light_controller_B.set_brightness(75)
    time.sleep(0.5)

    print("\n--- 路燈 SL-003 模擬故障 ---")
    light_controller_C.simulate_fault()
    time.sleep(0.5)

    print("\n--- 路燈 SL-001 調整亮度 ---")
    light_controller_A.set_brightness(100)
    time.sleep(0.5)

    print("\n--- 路燈 SL-002 關閉 ---")
    light_controller_B.turn_off()
    time.sleep(0.5)

    print("\n--- 警報系統不再監控 SL-003 ---")
    light_controller_C.detach(alert_system)

    print("\n--- 路燈 SL-003 故障排除 (假設修復後重新開啟) ---")
    # 實際中故障排除可能需要額外邏輯，這裡簡化為直接開啟
    light_controller_C.turn_on(90)
    time.sleep(0.5)

    print("\n===== 範例結束 =====")